[receiver/awss3receiver]: Add ingest progress notifications via OpAMP

adcharre opened the pull request:
**Description**: Add support for monitoring the progress of ingesting data from an S3 bucket via OpAMP custom messages.
This change would allow an appropriate UI to display the state of ingested. 

This is a re-working of #33047 to use OTLP structures instead of a custom JSON object following discussions here: https://cloud-native.slack.com/archives/C02J58HR58R/p1715714974462069

**Link to tracking Issue**: #30750

**Testing**: Additional unit tests added. Full stack testing using a built collector and the example OpAMP server provided in the PR.

**Documentation**: README.md updated to include details of the configuration changes and details of the OpAMP Custom Capability and message format.

adcharre committed:
[receiver/awss3receiver]:  Add progress notifications

Use OpAMP custom messages to report progress of start/end time based ingest from an S3 bucket.

adcharre committed:
Add example usage of notifications and OPAMP

adcharre committed:
Update go.mod

adcharre committed:
Add makefile to notifications_example

adcharre committed:
Move example to examples directory

adcharre committed:
Fix go version

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Fix unit tests

adcharre committed:
Correct package name

adcharre committed:
Change type to enhancement

adcharre committed:
Fix checkapi issue

adcharre committed:
go mod tidy

adcharre committed:
gci s3reader_test.go

adcharre committed:
Fix race condition in notifications test

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update to use OTLP Logs to send status

atoulme commented:
Please fix the conflict. Please have @evan-bradley and @tigrannajaryan review as this involves opAmp.

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Fix lint issues

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Fix deps

adcharre committed:
Lint fixes

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update receiver.go

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

evan-bradley commented on receiver/awss3receiver/notifications.go:
```diff
@@ -0,0 +1,105 @@
+	case err == nil:
+		break
+	case errors.Is(err, types.ErrCustomMessagePending):
+		<-sendingChan
```
How do retries work here? Right now this will just exit the function after the channel emits.

evan-bradley commented on receiver/awss3receiver/notifications_test.go:
```diff
@@ -0,0 +1,212 @@
+	if m.unregisterCalled {
+		return nil, fmt.Errorf("unregister called")
+	}
+	m.sentMessages = append(m.sentMessages, customMessage{messageType: messageType, message: message})
+	if m.shouldReturnPending {
+		return m.pendingChannel, types.ErrCustomMessagePending
+	}
```
```suggestion
	if m.shouldReturnPending {
		return m.pendingChannel, types.ErrCustomMessagePending
	}
	m.sentMessages = append(m.sentMessages, customMessage{messageType: messageType, message: message})
```
I think we need to not record a message as sent if one is pending. The [client implementation](https://github.com/open-telemetry/opamp-go/blob/main/client/internal/clientcommon.go#L403) doesn't allow sending messages no messages are currently queued.

evan-bradley commented on receiver/awss3receiver/config.go:
```diff
@@ -26,17 +26,22 @@ type S3DownloaderConfig struct {
 }
 
+type Notifications struct {
+	OpAMP *component.ID `mapstructure:"opamp"`
```
```suggestion
	OpAMP *component.ID `mapstructure:"opampextension"`
```
We should probably call out that we want the component ID of an extension here, since `opamp` could imply something like an endpoint.

evan-bradley commented on receiver/awss3receiver/s3reader.go:
```diff
@@ -64,15 +67,45 @@ func (s3Reader *s3Reader) readAll(ctx context.Context, telemetryType string, dat
 		case <-ctx.Done():
+			s3Reader.sendStatus(ctx, StatusNotification{
+				TelemetryType: telemetryType,
+				IngestStatus:  IngestStatusCompleted,
```
If the context finishes, would this be considered "completed" or "failed"? I suspect a context finish would indicate that something has gone wrong elsewhere and the transfer cannot be successfully completed.

evan-bradley replied:
@adcharre This isn't critical since completed/failed are "done" either way, but what are your thoughts on this?

adcharre replied:
I think this probably should be failed and also the return value should be the context error.

evan-bradley reviewed (commented):
Sorry about the long delay for the review.

adcharre commented:
> Sorry about the long delay for the review.

Thanks for the feedback and no worries I've been delayed in getting back to this - I am looking at the feedback, I've just not had the time I'd like to concentrate on this at the moment.

adcharre committed:
Unexport StatusNotification

adcharre commented on receiver/awss3receiver/notifications.go:
```diff
@@ -0,0 +1,105 @@
+	"github.com/open-telemetry/opentelemetry-collector-contrib/extension/opampcustommessages"
+)
+
+const (
```
I've un-exported StatusNotification but do you think the constants should stay exported to allow use by those wanting to process the messages?

evan-bradley replied:
That makes sense to me, let's keep exporting those.

adcharre committed:
Update receiver/awss3receiver/notifications_test.go

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

adcharre committed:
Update receiver/awss3receiver/README.md

Co-authored-by: Evan Bradley <11745660+evan-bradley@users.noreply.github.com>

adcharre committed:
Rename opamp config setting to opampextension

and add test

adcharre committed:
Merge branch 'awss3receiver_notifications' of https://github.com/adcharre/opentelemetry-collector-contrib into awss3receiver_notifications

adcharre committed:
Add retry logic

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Fix race conditions in notification tests

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update examples go.mod

adcharre committed:
Update go.sum

adcharre commented:
@evan-bradley I think that should address all of your comments, I'd completely missed the pending response meaning a retry of the send was required thanks again for the review!

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
On context done send failure notification

evan-bradley commented:
There are two go.mod-related build failures. I think one may be fixed if you move the example to an internal directory. For the other one, you will have to update the s3 receiver's version of pdata. A merge may do it, otherwise I would just check the error in the failed job.

evan-bradley commented:
@BinaryFissionGames It would be great if you could give this a look.

adcharre commented on receiver/awss3receiver/examples/notifications/go.mod:
Would this actually better moved under `/examples/awss3receiver/notifications` ? Moving under the internal directory just makes it even more difficult to find ...

adcharre committed:
Update example go.mod

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update example go.mod

adcharre committed:
Move the notifications example to the examples/ directory

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update examples go.mod

adcharre commented:
@evan-bradley  - I've moved the example out to the `examples/` directory which hopefully helps. Are you now happy enough to approve from your side?

BinaryFissionGames commented on receiver/awss3receiver/README.md:
```diff
@@ -32,6 +32,8 @@ The following exporter configuration parameters are supported.
 | `extension`             | Extension to use for decoding a key with a matching suffix.                                                                                |             | Required |
 | `suffix`                | Key suffix to match against.                                                                                                               |             | Required |
+| `notifications:`        |                                                                                                                                            |             |          |
+| `opamp`                 | Name of the OpAMP Extension to use to send ingest progress notifications.                                                               |             |          |
```
```suggestion
| `opampextension`                 | Name of the OpAMP Extension to use to send ingest progress notifications.                                                               |             |          |
```

adcharre replied:
Thanks updated!

BinaryFissionGames commented on receiver/awss3receiver/README.md:
```diff
@@ -67,4 +69,25 @@ receivers:
+| `telemetry_type`  | The type of telemetry being ingested. One of "traces", "metrics", or "logs".    |
+| `ingest_status`   | The status of the data ingestion. One of "ingesting", "failed", or "completed". |
+| `start_time`      | The time to start retrieving data in RFC3339 format.                            |
+| `stop_time`       | The time to stop retrieving data in RFC3339 format.                             |
```
Looks like this is `end_time` in code.
```suggestion
| `end_time`       | The time to stop retrieving data in RFC3339 format.                             |
```

adcharre replied:
Thanks updated!

BinaryFissionGames reviewed (commented):
Sorry for taking so long to look at this. It looks good, just noticed a couple discrepancies in the README vs the code.

adcharre committed:
Delete Makefile

adcharre committed:
Remove example

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update README.md

adcharre commented:
@evan-bradley / @BinaryFissionGames - thanks for the reviews. I've updated with all the comments now and removed the example for the moment as it's not clear to me where this should live - I may move this out into a separate repo.

tigrannajaryan commented on receiver/awss3receiver/notifications.go:
```diff
@@ -0,0 +1,114 @@
+	attributes := log.Attributes()
+	attributes.PutStr("telemetry_type", message.TelemetryType)
+	attributes.PutStr("ingest_status", message.IngestStatus)
+	attributes.PutStr("start_time", message.StartTime.Format(time.RFC3339))
```
In OTLP world it is common to use [uint64 nanoseconds](https://github.com/open-telemetry/opentelemetry-proto/blob/14afbd4e133ee8a8a5a9f7a0fd3a09d5a9456340/opentelemetry/proto/logs/v1/logs.proto#L140) for timestamps. Would it make sense to do the same here too instead of string-formatted times?

Also we have 2 native timestamp fields in LogRecord, perhaps use these for one of the 3 timestamps you need (instead of an attribute), if the semantics fit?

adcharre replied:
Thanks @tigrannajaryan - I've left start/end time as attributes and used the Timestamp (not observed timestamp) for the ingested time.

tigrannajaryan replied:
👍

adcharre committed:
Change notification time types to int64

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update notifications_test.go

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update README.md

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Update notifications_test.go

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre commented:
@atoulme / @evan-bradley  / @tigrannajaryan / @BinaryFissionGames  - Can I get another review please!

adcharre committed:
fix testifylint issues

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

atoulme commented on receiver/awss3receiver/notifications.go:
```diff
@@ -0,0 +1,116 @@
+}
+
+func (n *opampNotifier) Shutdown(_ context.Context) error {
+	n.handler.Unregister()
```
nil check on handler

adcharre replied:
I've updated the code and added a test case to cover this.

atoulme reviewed (approved):
Approved. One nit to secure the code a bit more, but not really needed, since you're checking for notifier itself in Shutdown.

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Add nil handler check

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Check handler != nil before shutdown

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications

adcharre committed:
Merge remote-tracking branch 'upstream/main' into awss3receiver_notifications
